crates/runtime-interface/src/traits.rs

```diff
@@ -26,11 +29,19 @@ pub trait MohoProgram {
     ///
     /// If this returns error, proving fails.
     // TODO make result type
-    fn process_transition(pre_state: &Self::State, inp: &Self::StepInput) -> Self::State;
+    fn process_transition(pre_state: &Self::State, inp: &Self::StepInput) -> Self::StepOutput;
+
+    /// Extracts the next inner verification key from a step’s output.
+    ///
+    /// # Returns
+    ///
+    /// - `Some(InnerVerificationKey)` if the inner verification key has been updated.
+    /// - `None` if there is no update to the inner verification key.
+    fn extract_next_vk(output: &Self::StepOutput) -> Option<InnerVerificationKey>;
 
-    /// Extracts the next vk from the state.
-    fn extract_next_vk(state: &Self::State) -> InnerVerificationKey;
+    /// Extracts the inner state after a transition from the step’s output.
+    fn extract_post_state(output: &Self::StepOutput) -> &Self::State;
 
-    /// Extracts the exported output from the state.
-    fn extract_export_state(state: &Self::State) -> ExportState;
+    /// Updates the exported state from the output.
+    fn update_export_state(export_state: &mut ExportState, output: &Self::StepOutput);
```

Comment:
Regarding the other comment, we can avoid including these as functions on the type if we provide more structure in what `process_transition` returns.